libtock: aes library syncronous

niklasad1 opened the pull request:
Hi,
I noticed the several other libraries provide synchronous system calls which is nice especially for aes.
Thus, I changed wrappers in libtock which makes it easier to use in my opinion.
If you agree with this I will make the same change for the temperature sensor for nrf51 as well.

--
Niklas A

brghena commented:
This is a good idea. I'd say the style here should change a little though.

If you take a look at the libtock driver for the [ADC](https://github.com/helena-project/tock/blob/master/userland/libtock/adc.h), I think the best way to structure this is to have functions that wrap the syscalls (which can be called directly) and synchronous wrappers of operations (usually a couple syscalls, at least a subscribe, allow, and command).

The current convention is to end the synchronous commands with `_sync`. I've also switched to using a result allocated on the stack rather than globally in the file. Example: https://github.com/helena-project/tock/blob/master/userland/libtock/adc.c#L233

brghena commented:
Thanks for those changes! Two remaining things:

1. We definitely want the wrappers around syscalls in addition to this. So there should be a function for each subscribe, allow, and command that's valid. Then the `_sync` functions should call combinations of those functions.

2. It would be great to add comments to this file documenting what each function does and what its arguments are.

Again, I think the ADC libtock files are a good example of this.

niklasad1 commented:
@brghena 
1. I thought that was already handled in the previous commit i.e., all wrappers were called _sync or am I missing something? However, I made some minor changes I got annoyed by my previous implementation, it makes no sense to make set_key in the capsule asynchronous with the client HIL. Thus, it is now synchronous in the kernel as well.
2. added comments let me know if it is good enough and my english is not to bad :)

brghena commented on userland/libtock/aes.h:
```diff
@@ -8,13 +8,45 @@ extern "C" {
+//
+// buf      - buffer to encrypt (currently max 128 bytes are supported)
+// buf_len  - length of the buffer to encrypt
+// ctr      - buffer with the initial counter
```
Is there an expected or maximum length for the initial counter buffer?

brghena commented on userland/libtock/aes.h:
```diff
@@ -8,13 +8,45 @@ extern "C" {
+//
+// buf      - buffer to decrypt (currently max 128 bytes are supported)
+// buf_len  - length of the buffer to decrypt
+// ctr      - buffer with the initial counter
```
Is there an expected or maximum length for the initial counter buffer?

niklasad1 replied:
16 bytes (one aes block), generally this is checked by the capsule but perhaps I should add a comment about that as well.

brghena commented on kernel/src/hil/symmetric_encryption.rs:
```diff
@@ -18,7 +18,7 @@ pub trait SymmetricEncryptionDriver {
     /// Configure encryption/decryption key
     /// assumes that key size is 16, 24 or 32 bytes
-    fn set_key(&self, key: &'static mut [u8], len: usize);
+    fn set_key(&self, key: &'static mut [u8], len: usize) -> &'static mut [u8];
```
Making `set_key` synchronous here is a great idea.

brghena commented on userland/libtock/aes.c:
```diff
@@ -1,37 +1,100 @@
-  return command(AES_DRIVER, AES_DEC, 0);
+  if (err < SUCCESS) return err;
+
+  err = command(AES_DRIVER, AES_DEC, 0);
```
So what I meant by wrappers for system calls is that you should have a function like:

```
int aes128_decrypt (void) {
  return command(AES_DRIVER, AES_DEC, 0);
}
```

And then your `aes128_decrypt_ctr_sync` function here would call:
```
  err = aes128_decrypt();
```

The goal of structuring drivers in this way is that users could either call the system calls directly and handle the asynchronous stuff themselves, or they can just call `aes128_decrypt_ctr_sync` and have the operation performed synchronously. Furthermore, if they have to use the functions asynchronously (for some app-specific reason) they can look at the body of `aes128_decrypt_ctr_sync` to figure out what functions they will need to call.

niklasad1 replied:
I see, thanks

niklasad1 commented:
I think, the 50 changed files are GitHub bug because other commits have been merged after my PR.
I can withdraw the PR and submit again if you would like that?

(acc. to git diff 6 files are changed)

brghena commented:
I fixed it. In the future, what you need to do is first pull your master branch up to the Tock master branch, and then rebase your branch upon it (`git rebase origin/master`). If master ever has new commits, doing a rebase is the best way to move your pull request above them.

brghena commented:
Oh, this will definitely make your local branch angry. What you have to do is:

```
git checkout master
git branch -D aes_library_refactor
git pull
git checkout aes_library_refactor
```

niklasad1 commented:
At least I added wrappers (not sure if they are good ones didn't become as nice as your example) for the sync and async versions. 

Further,  aes_encrypt() and aes_decrypt() are perhaps not "good" names because these can probably confuse people but those are not supposed to be used by "application developer" (only internally in libtock).

I added a very "rough" test for the async versions as well basically prints out the contents of the buffer

brghena commented on userland/libtock/aes.c:
```diff
@@ -1,37 +1,165 @@
+// Internal function to trigger encryption operation. Note that this don't
+// work by itself aes128_set_data() and aes128_set_ctr() must be called first
+// FIXME: better name?
+int aes128_encrypt(void) {
```
I'm happy with them being named whatever you think is correct. I agree with `aes128_encrypt` being a little confusing. Maybe `aes128_start_encryption`?

brghena replied:
(also `don't` should be `doesn't`)

brghena commented on userland/libtock/aes.c:
```diff
@@ -1,37 +1,165 @@
-  }
-  return command(AES_DRIVER, AES_KEY, 0);
+// Internal function to configure a payload to encrypt or decrypt
+int aes128_set_data(const unsigned char *data, unsigned char len) {
```
Exactly what I was thinking. Thank you!

brghena commented on userland/tools/uncrustify/uncrustify.cfg:
```diff
@@ -1,5 +1,111 @@
 # Uncrustify-0.65_f
 
+# ====================================================================
```
Problems with history again. If you check the list of commits for this PR, you'll see that it includes one of Pat's, which is why it thinks this file is modified. Also, this PR can't even be merged right now because it conflicts with master, probably the same reason. I'm not sure what commands you're running to merge master, but something is wrong with them.

You should pull your master up to date and `git rebase` on top of it.

niklasad1 replied:
yes, I'm generally too sloppy and perform rebasing as the last step but I guess I should be rebasing all the time i.e., I have to make it a habit!!!

brghena commented:
I'll merge once the Travis checks pass, if you're good to go

niklasad1 commented:
Yes, good enough for me.
Thanks for the help. I will try to rebase every commit in the future.
